library/waku_thread/inter_thread_communication/waku_thread_request.nim

```diff
@@ -0,0 +1,55 @@
+
+## This file contains the base message request type that will be handled.
+## The requests are created by the main thread and processed by
+## the Waku Thread.
+
+import
+  std/json,
+  stew/results
+import
+  chronos
+import
+  ../../../waku/node/waku_node,
+  ./requests/node_lifecycle_request,
+  ./requests/peer_manager_request,
+  ./requests/protocols/relay_request
+
+type
+  RequestType* {.pure.} = enum
+    LIFECYCLE,
+    PEER_MANAGER,
+    RELAY,
+
+type
+  InterThreadRequest* = object
+    reqType: RequestType
+    reqContent: pointer
+
+proc new*(T: type InterThreadRequest,
+          reqType: RequestType,
+          reqContent: pointer): ptr InterThreadRequest =
+  var ret = cast[ptr InterThreadRequest](
+                                      allocShared0(sizeof(InterThreadRequest)))
+  ret[].reqType = reqType
+  ret[].reqContent = reqContent
+  return ret
+
+proc process*(T: type InterThreadRequest,
+              request: ptr InterThreadRequest,
+              node: ptr WakuNode):
+              Result[string, string] =
+  ## Processes the request and deallocates its memory
+  defer: deallocShared(request)
+
+  echo "Request received: " & $request[].reqType
+
+  case request[].reqType
+    of LIFECYCLE:
+      waitFor cast[ptr NodeLifecycleRequest](request[].reqContent).process(node)
```

Comment:
Well, in this case, the `waitFor` is "sync", and it keeps there making the dispatcher progress until the `process` is done. The `reqContent` pointer will only be deallocated by the `process` proc when it completes.

Reply:
Well, afaik, the `waitFor` is "sync", and it keeps there making the dispatcher progress until the `process` is done. The `reqContent` pointer will only be deallocated by the `process` proc when it completes.

Reply:
`waitFor` turns async code into sync code indeed - but a problem with this setup is that the async loop will only be running while `waitFor` is running - this means that it relies on a stead flow of requests in order to process networking buffers and other waku activity (timers, etc).

To solve this, `process` itself must be `async` and this should be turned into `await process` - the "main" loop then needs to call `poll` / `runForever`

Reply:
right - this runs the risk of getting stuck in `poll` because `poll` itself will block until there is activity, and if there is no activity it will simply block forever - in fact, I'm not sure how this works at all - it should get stuck there and never perform any loop iterations - I'm probably missing some other detail which wakes up poll but this looks like a significant risk with the setup.

There are two ways to solve this: introducing a timer / `sleepAsync` (`waitFor sleepAsync(1.millis)` instead of `poll`) or using a [`ThreadSignal`]. Introducing a timer is the easier way to make this code correct.

Using a signal has the advantage of using fewer resources, but is slightly more difficult to implement. I recommend leaving it for a separate PR - the way to use that is to create a signal for every channel (this is the way channels are normally implemented, with a "notification mechanism"), and every time an item is added to the channel, the signal is fired - here's a simple example: https://gist.github.com/arnetheduck/b6a7ac8f4b85490d26d464674e09d57d#file-threadsynctask-nim

Reply:
"in general" I would look for ways / strive to avoid calling `waitFor` in "inner" proc's and rather structure the code in such a way that the `waitFor` happens only at the "outer" layer - the waitFor strategy is not incorrect, but it's unusual and runs the risk of accidentally ending up with nested `waitFor` calls (calling `waitFor` from an `async` function) which is not supported.

Reply:
> "in general" I would look for ways / strive to avoid calling `waitFor` in "inner" proc's...

Okay perfect. I enhanced that in my last commit.

Reply:
> right - this runs the risk of getting stuck in `poll` because `poll` itself will block...

Ok thanks! We'll apply `waitFor sleepAsync(1)` for now and will implement the enhancement with signal in further PRs.